src/lib/storage/segment_access_statistics.hpp

```diff
@@ -0,0 +1,379 @@
+
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+  static std::chrono::time_point<std::chrono::steady_clock> start_time;
+  static std::chrono::duration<double, std::milli> interval;
+
+ private:
+
+  struct Counter {
+    uint64_t timestamp;
+    std::array<std::atomic_uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> counter;
+
+    Counter(uint64_t timestamp) : timestamp{timestamp}, counter{} {};
+    Counter() : Counter(0ul) {}
+    Counter(const Counter& other) {
+      timestamp = other.timestamp;
+      for (auto type = 0ul; type < counter.size(); ++type) {
+        counter[type] = other.counter[type].load();
+      }
+    }
+    Counter(Counter&& other) noexcept {
+      timestamp = other.timestamp;
+      for (auto type = 0ul; type < counter.size(); ++type) {
+        counter[type] = other.counter[type].load();
+      }
+    }
+  };
+
+//  struct Counter {
+//    uint64_t timestamp;
+//    std::array<uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> counter;
+//
+//    Counter(uint64_t timestamp) : timestamp{timestamp}, counter{} {};
+//    Counter() : Counter(0ul) {}
+//  };
+
+  pmr_concurrent_vector<Counter> _counters;
+  Counter& _counter();
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+class NonLockingStrategy {
+ public:
+  uint64_t count(SegmentAccessType type) const;
+
+  void reset();
+
+  void increase(SegmentAccessType type, uint64_t count);
+
+  static std::string header();
+
+  std::vector<std::string> to_string() const;
+
+ private:
+  std::array<uint64_t, static_cast<uint32_t>(SegmentAccessType::Count)> _count;
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+template<class DataAccessStrategy>
+class BulkCountingStrategy {
+ public:
+  explicit BulkCountingStrategy(DataAccessStrategy& data_access_strategy)
+    : _data_access_strategy {data_access_strategy} {}
+
+  void on_iterator_create(uint64_t count) {
+     _data_access_strategy.increase(SegmentAccessType::IteratorCreate, 1);
+     _data_access_strategy.increase(SegmentAccessType::IteratorSeqAccess, count);
+  }
+
+  void on_iterator_create(const std::shared_ptr<const PosList>& positions) {
+    // hier muss geprüft werden, um was für einen Zugriff es sich handelt
+    _data_access_strategy.increase(SegmentAccessType::IteratorCreate, 1);
+    _data_access_strategy.increase(SegmentAccessTypeTools::iterator_access_pattern(positions), positions->size());
+  }
+
+  void on_iterator_dereference(uint64_t count) {}
+
+  void on_iterator_dereference(uint64_t count, ChunkOffset chunk_offset) {}
+
+  void on_accessor_create(uint64_t count) {
+    _data_access_strategy.increase(SegmentAccessType::AccessorCreate, 1);
+  }
```

Comment:
Do we plan on using the information on how many accessors were created? For now, tracking more than necessary is fine, but once we bring this in, we should have a look at what is actually needed.

Reply:
I don't think so. In case of the iterators it might make sense though.